### PR TITLE
cgroup: Imply systemd.unified_cgroup_hierarchy=1 on cgroup_no_v1=all

### PR DESCRIPTION
**Pull request by cdown**
cgroup_no_v1=all doesn't make a whole lot of sense with legacy hierarchy (where we use v1 hierarchy for everything), or hybrid hierarchy (where we still use v1 hierarchy for resource control).

Right now we have to tell people to add both cgroup_no_v1=all and systemd.unified_cgroup_hierarchy=1 to get the desired behaviour, however in reality it's hard to imagine any situation where someone passes cgroup_no_v1=all but *doesn't* want to use the unified cgroup hierarchy.

Make it so that cgroup_no_v1=all produces intuitive behaviour in systemd by default, although it can still be disabled by passing systemd.unified_cgroup_hierarchy=0 explicitly.

### CODE REVIEW

**Comment by cdown** [comment]
arm64 is just hung due to some flaky stuff, this is ready for review.

File: src/test/test-cgroup-util.c [hunk 1]
```diff
@@ -369,6 +369,17 @@ static void test_is_wanted(void) {
...
         test_is_wanted_print(false);
+
+        /* cgroup_no_v1=all implies systemd.unified_cgroup_hierarchy=1, unless
+         * otherwise explicitly specified. */
+        assert_se(setenv("SYSTEMD_PROC_CMDLINE",
+                         "cgroup_no_v1=all", 1) >= 0);
+        test_is_wanted_print(false);
+
+        assert_se(setenv("SYSTEMD_PROC_CMDLINE",
+                         "cgroup_no_v1=all"
```
Comment (keszybz) [inline src/test/test-cgroup-util.c]: Hmm, those strings will be concatenated by the compiler, so this test isn't working really.

File: src/basic/cgroup-util.c [hunk 2]
```diff
@@ -2716,11 +2717,20 @@ bool cg_is_unified_wanted(void) {
...
+        wanted = is_default;
+        r = proc_cmdline_get_key("cgroup_no_v1", 0, &c);
+        if (r > 0 && streq_ptr(c, "all"))
+                wanted = true;
+
+        /* ...but, let's see what the kernel command line has to say. Maybe
+         * we were explicitly passed systemd.unified_cgroup_hierarchy. */
         r = proc_cmdline_get_bool("systemd.unified_cgroup_hierarchy", &b);
+        if (r > 0)
+                wanted = b;
```
Comment (keszybz) [inline src/basic/cgroup-util.c]: This should be `r >= 0`, no?

Comment (cdown) [reply src/basic/cgroup-util.c]: I don't think so? `proc_cmdline_get_bool` returns 0 on "not found", as passed through from `proc_cmdline_get_key`

**Comment by poettering** [comment]
lgtm

File: src/basic/cgroup-util.c (hunk 2 above)
Comment (keszybz) [reply src/basic/cgroup-util.c]: Right.

**Comment by keszybz** [comment]
Hmm,
```
-- test_is_wanted_print --
cmdline: systemd.unified_cgroup_hierarchy=0
is_unified_wanted() → yes
is_hybrid_wanted() → no
is_legacy_wanted() → no
 
-- test_is_wanted_print --
cmdline: systemd.unified_cgroup_hierarchy=0 systemd.legacy_systemd_cgroup_controller
is_unified_wanted() → yes
is_hybrid_wanted() → no
is_legacy_wanted() → no
 
-- test_is_wanted_print --
cmdline: systemd.unified_cgroup_hierarchy=0 systemd.legacy_systemd_cgroup_controller=0
is_unified_wanted() → yes
is_hybrid_wanted() → no
is_legacy_wanted() → no
 
-- test_is_wanted_print --
cmdline: cgroup_no_v1=all
is_unified_wanted() → yes
is_hybrid_wanted() → no
is_legacy_wanted() → no
 
-- test_is_wanted_print --
cmdline: cgroup_no_v1=all systemd.unified_cgroup_hierarchy=0
is_unified_wanted() → yes
is_hybrid_wanted() → no
is_legacy_wanted() → no
```
The result was similar before this PR, but getting unified hierarchy after I explicitly said I don't want it, is at least surprising.

**Comment by keszybz** [comment]
Yeah, what the man page says describes different behaviour and makes much more sense:
> systemd.unified_cgroup_hierarchy
           When specified without an argument or with a true argument, enables the
           usage of unified cgroup hierarchy[9] (a.k.a. cgroups-v2). When specified
           with a false argument, fall back to hybrid or full legacy cgroup hierarchy.

**Comment by cdown** [comment]
> The result was similar before this PR, but getting unified hierarchy after I explicitly said I don't want it, is at least surprising.

Yeah, I wasn't sure what to make of these tests, but it shows the same before and after this PR. I wonder if we don't invalidate the cache properly?

I manually tested on a virtual machine and it works fine anyway, so meh.

**Comment by keszybz** [comment]
I wrote a test for this locally, just to make sure, and it all seems OK. But since it refactors the code a bit, I'll submit it after v240.

**Pull request merged**